### PR TITLE
Add more information to timeline-related APIs

### LINKED ISSUE #1488: Better current status information
While debugging an issue in staging, it was hard to tell what the current status of a timeline in the pageserver was.

Let's add a new HTTP API call that returns the following information for every timeline in the pageserver:

- loaded locally or remote-only?
- state: Active, Idle, Stopping (from TenantState in tenant_mgr.rs)
- wal_producer_connstr, from last callmemaybe
- WAL receiver thread ID
- Current WAL receiver's libpq connection's remote IP / port
- timestamp of last message received by WAL receiver msg 
- LSN of last received piece of WAL
- last_record_lsn and prev_record_lsn
- disk_consistent_lsn
- last GC lsn

Would be nice to get similar status information from each safekeeper, too.

### PR DESCRIPTION
**Pull request by aome510**
Resolves #1488.

- implemented `GET tenant/:tenant_id/timeline/:timeline_id/wal_receiver` endpoint
- returned `thread_id` in `thread_mgr::spawn` 
- added `latest_gc_cutoff_lsn` field to `LocalTimelineInfo` struct

### CODE REVIEW

**Comment by aome510** [comment]
The requirement copied from #1488 issue description:

- [x] loaded locally or remote-only?, last_record_lsn, prev_record_lsn, disk_consistent_lsn, last GC lsn
   + already included in the response of `GET tenant/:tenant_id/timeline/:timeline_id` API
- [x] state: Active, Idle, Stopping (from TenantState in tenant_mgr.rs)
  + already included in the response of `GET tenant/:tenant_id` API
- [x] Current WAL receiver's libpq connection's remote IP / port
  + can be determined from wal_producer_connstr
- [x] wal_producer_connstr , WAL receiver thread ID, timestamp of last message received by WAL receiver msg, LSN of last received piece of WAL
   + included in the response of `GET tenant/:tenant_id/timeline/:timeline_id/wal_receiver` API

File: pageserver/src/walreceiver.rs [hunk 1]
```diff
@@ -118,7 +128,14 @@ fn thread_main(conf: &'static PageServerConf, tenant_id: ZTenantId, timeline_id:
     info!("WAL receiver thread started");
 
     // Look up the current WAL producer address
-    let wal_producer_connstr = get_wal_producer_connstr(tenant_id, timeline_id);
+    let wal_producer_connstr = get_wal_receiver_entry(tenant_id, timeline_id)
+        .unwrap_or_else(|| {
+            panic!(
```
Comment (hlinnaka) [inline pageserver/src/walreceiver.rs:133]: Let's change this into an Err return. (Yeah, it was a panic before, via unwrap(), but a softer error seems more appropriate here.)

Something like `.ok_or_else(anyhow!("..."))?` would work I think

File: pageserver/src/walreceiver.rs [hunk 2]
```diff
@@ -318,6 +335,26 @@ fn walreceiver_main(
             let apply_lsn = u64::from(timeline_remote_consistent_lsn);
             let ts = SystemTime::now();
 
+            // Update the current WAL receiver's data stored inside the global hash table `WAL_RECEIVERS`
+            {
+                let mut receivers = WAL_RECEIVERS.lock().unwrap();
+                let entry = receivers
+                    .get_mut(&(tenant_id, timeline_id))
+                    .unwrap_or_else(|| {
+                        panic!(
```
Comment (hlinnaka) [inline pageserver/src/walreceiver.rs:344]: `panic!` -> `?` here too

File: pageserver/src/walreceiver.rs (hunk 1 above)
Comment (aome510) [reply pageserver/src/walreceiver.rs]: updated! The function doesn't return an error, so I need to change the code differently with `return`

File: pageserver/src/tenant_mgr.rs [hunk 3]
```diff
@@ -257,6 +257,7 @@ pub fn activate_tenant(tenant_id: ZTenantId) -> anyhow::Result<()> {
                 false,
                 move || crate::tenant_threads::gc_loop(tenant_id),
             )
+            .map(|_| ())
```
Comment (SomeoneToIgnore) [inline pageserver/src/tenant_mgr.rs:260]: NIT: is it needed?

File: pageserver/src/http/openapi_spec.yml [hunk 4]
```diff
@@ -520,6 +567,21 @@ components:
...
+        - wal_producer_connstr
+      properties:
+        thread_id:
+          type: integer
+        wal_producer_connstr:
+          type: string
+        last_received_msg_lsn:
+          type: string
+          format: hex
+        last_received_msg_ts:
```
Comment (SomeoneToIgnore) [inline pageserver/src/http/openapi_spec.yml:583]: > `_ts`

Still not sure what does this stand for, might make sense to add a doc comment (`///`) on a corresponding Rust struct field or a `description: ` field here or both.

**Review by SomeoneToIgnore — APPROVED** [review]
Couple of nits, LTGM generally.

File: test_runner/batch_others/test_pageserver_api.py [hunk 5]
```diff
@@ -12,6 +19,25 @@ def test_pageserver_init_node_id(zenith_env_builder: ZenithEnvBuilder):
         env.pageserver.start(overrides=['--pageserver-config-override=id=10'])
 
 
+def create_tenant_and_root_branch(
```
[the remark this answers is not in the document]
Comment (aome510) [reply test_runner/batch_others/test_pageserver_api.py:22]: I was wondering the same thing before :). `client.tenant_create` just makes a `/tenant/:id/create` API call and it doesn't create an initial timeline for the client. If we change that to `zenith_cli.tenant_create`, then yes we don't need this helper function anymore after #1689

File: pageserver/src/tenant_mgr.rs (hunk 3 above)
Comment (aome510) [reply pageserver/src/tenant_mgr.rs:260]: One change is made in this PR to return `thread_id` upon spawning new thread. the `activate_tenant` requires the return signature to `Result<()>`, so this mapping is needed. Will add a comment to make this more clear

File: test_runner/batch_others/test_pageserver_api.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (SomeoneToIgnore) [reply test_runner/batch_others/test_pageserver_api.py:22]: Indeed.

We have something similar with etcd management now: we have to keep the code for managing it both in python and cli (via control_plane) code.

If you feel motivated enough, you can check on the possibility to refactor the Python code to use `neon_base` (maybe some extra methods in the cli would be needed).

But that's a separate thing to consider, this function seems good enough for now.

File: pageserver/src/tenant_mgr.rs (hunk 3 above)
Comment (SomeoneToIgnore) [reply pageserver/src/tenant_mgr.rs:260]: Or you could give an omitted parameter a readable name?
```suggestion
            .map(|_thread_id | ())
```

File: test_runner/batch_others/test_pageserver_api.py [hunk 6]
```diff
@@ -48,6 +54,40 @@ def check_client(client: ZenithPageserverHttpClient, initial_tenant: UUID):
...
+    assert list(res.keys()) == [
+        "thread_id",
+        "wal_producer_connstr",
+        "last_received_msg_lsn",
+        "last_received_msg_ts",
+    ]
+
+    # make a DB modification then expect getting a new WAL receiver's data
+    cur = pg.connect().cursor()
+    cur.execute("CREATE TABLE t(key int primary key, value text)")
```
Comment (LizardWizzard) [inline test_runner/batch_others/test_pageserver_api.py:86]: nit: can be `pg.safe_psql`

**Review by LizardWizzard — APPROVED** [review]
Looks good to me. I was thinking about adding that directly to timeline detail handler, but no objections to have a separate handler for that

**Pull request merged**